File: clang/test/CodeGenObjC/noescape.m
// RUN: %clang_cc1 -fblocks -emit-llvm -o - %s | FileCheck %s

typedef void (^BlockTy)(void);

union U {
  int *i;
  long long *ll;
} __attribute__((transparent_union));

void noescapeFunc0(id, __attribute__((noescape)) BlockTy);
void noescapeFunc1(__attribute__((noescape)) int *);
void noescapeFunc2(__attribute__((noescape)) id);
void noescapeFunc3(__attribute__((noescape)) union U);

// CHECK-LABEL: define void @test0(
// CHECK: call void @noescapeFunc0({{.*}}, [[TY0:.*]] nocapture {{.*}})
// CHECK: declare void @noescapeFunc0(i8*, [[TY0]] nocapture)
void test0(BlockTy b) {
  noescapeFunc0(0, b);
}

// CHECK-LABEL: define void @test1(
// CHECK: call void @noescapeFunc1([[TY1:.*]] nocapture {{.*}})
// CHECK: declare void @noescapeFunc1([[TY1]] nocapture)
void test1(int *i) {
  noescapeFunc1(i);
}

// CHECK-LABEL: define void @test2(
// CHECK: call void @noescapeFunc2([[TY2:.*]] nocapture {{.*}})
// CHECK: declare void @noescapeFunc2([[TY2]] nocapture)
void test2(id i) {
  noescapeFunc2(i);
}

// CHECK-LABEL: define void @test3(
// CHECK: call void @noescapeFunc3([[TY3:.*]] nocapture {{.*}})
// CHECK: declare void @noescapeFunc3([[TY3]] nocapture)
void test3(union U u) {
  noescapeFunc3(u);
}

File: clang/test/CodeGenCXX/noescape.cpp
// RUN: %clang_cc1 -std=c++11 -emit-llvm -o - %s | FileCheck %s

struct S {
  int a[4];
  S(int *, int * __attribute__((noescape)));
  S &operator=(int * __attribute__((noescape)));
  void m0(int *, int * __attribute__((noescape)));
  virtual void vm1(int *, int * __attribute__((noescape)));
};

// CHECK: define void @_ZN1SC2EPiS0_U8noescape(%struct.S* {{.*}}, {{.*}}, {{.*}} nocapture)
// CHECK: define void @_ZN1SC1EPiS0_U8noescape(%struct.S* {{.*}}, {{.*}}, {{.*}} nocapture) {{.*}} {
// CHECK: call void @_ZN1SC2EPiS0_U8noescape(%struct.S* {{.*}}, {{.*}}, {{.*}} nocapture {{.*}})

S::S(int *, int * __attribute__((noescape))) {}

// CHECK: define {{.*}} %struct.S* @_ZN1SaSEPiU8noescape(%struct.S* {{.*}}, {{.*}} nocapture)
S &S::operator=(int * __attribute__((noescape))) { return *this; }

// CHECK: define void @_ZN1S2m0EPiS0_U8noescape(%struct.S* {{.*}}, {{.*}} nocapture)
void S::m0(int *, int * __attribute__((noescape))) {}

// CHECK: define void @_ZN1S3vm1EPiS0_U8noescape(%struct.S* {{.*}}, {{.*}} nocapture)
void S::vm1(int *, int * __attribute__((noescape))) {}

// CHECK-LABEL: define void @_Z5test0P1SPiS1_(
// CHECK: call void @_ZN1SC1EPiS0_U8noescape(%struct.S* {{.*}}, {{.*}}, {{.*}} nocapture {{.*}})
// CHECK: call {{.*}} %struct.S* @_ZN1SaSEPiU8noescape(%struct.S* {{.*}}, {{.*}} nocapture {{.*}})
// CHECK: call void @_ZN1S2m0EPiS0_U8noescape(%struct.S* {{.*}}, {{.*}}, {{.*}} nocapture {{.*}})
// CHECK: call void {{.*}}(%struct.S* {{.*}}, {{.*}}, {{.*}} nocapture {{.*}})
void test0(S *s, int *p0, int *p1) {
  S t(p0, p1);
  t = p1;
  s->m0(p0, p1);
  s->vm1(p0, p1);
}

namespace std {
  typedef decltype(sizeof(0)) size_t;
}

// CHECK: define {{.*}} @_ZnwmPvU8noescape({{.*}}, {{.*}} nocapture {{.*}})
void *operator new(std::size_t, void * __attribute__((noescape)) p) {
  return p;
}

// CHECK-LABEL: define i8* @_Z5test1Pv(
// CHECK : %call = call {{.*}} @_ZnwmPv({{.*}}, {{.*}} nocapture {{.*}})
void *test1(void *p0) {
  return ::operator new(16, p0);
}

// CHECK-LABEL: define void @_Z5test2PiS_(
// CHECK: call void @"_ZZ5test2PiS_ENK3$_0clES_S_U8noescape"({{.*}}, {{.*}}, {{.*}} nocapture {{.*}})
// CHECK: define internal void @"_ZZ5test2PiS_ENK3$_0clES_S_U8noescape"({{.*}}, {{.*}}, {{.*}} nocapture)
void test2(int *p0, int *p1) {
  auto t = [](int *, int * __attribute__((noescape))){};
  t(p0, p1);
}
